crates/polars-core/src/chunked_array/logical/categorical/merge.rs

```diff
@@ -102,52 +99,95 @@ impl RevMapMerger {
         Ok(())
     }
 
-    pub(crate) fn finish(self) -> Arc<RevMapping> {
+    pub fn finish(self) -> Arc<RevMapping> {
         match self.state {
             None => self.original,
             Some(state) => {
-                let new_rev = RevMapping::Global(state.map, state.slots.into(), self.id.unwrap());
+                let new_rev = RevMapping::Global(state.map, state.slots.into(), self.id);
                 Arc::new(new_rev)
             },
         }
     }
 }
 
-pub(crate) fn merge_rev_map(
-    left: &Arc<RevMapping>,
-    right: &Arc<RevMapping>,
-) -> PolarsResult<Arc<RevMapping>> {
-    match (&**left, &**right) {
-        (RevMapping::Global(_, _, _), RevMapping::Global(_, _, _)) => {
-            let mut merger = RevMapMerger::new(left.clone());
-            merger.merge_map(right)?;
-            Ok(merger.finish())
-        },
-        (RevMapping::Local(arr_l), RevMapping::Local(arr_r)) => {
-            // they are from the same source, just clone
-            if std::ptr::eq(arr_l, arr_r) {
-                return Ok(left.clone());
-            }
-
-            let arr = arrow::compute::concatenate::concatenate(&[arr_l, arr_r]).unwrap();
-            let arr = arr
-                .as_any()
-                .downcast_ref::<Utf8Array<i64>>()
-                .unwrap()
-                .clone();
-
-            Ok(Arc::new(RevMapping::Local(arr)))
-        },
-        _ => polars_bail!(
-            ComputeError:
-            "unable to merge categorical under a global string cache with a non-cached one"
-        ),
+pub fn merge_local_rhs_categorical<'a>(
+    categories: &'a Utf8Array<i64>,
+    ca_right: &'a CategoricalChunked,
+) -> Result<(impl Iterator<Item = Option<u32>> + 'a, Arc<RevMapping>), PolarsError> {
+    // Counterpart of the GlobalRevmapMerger.
+    // In case of local categorical we also need to change the physicals not only the revmap
+
+    let RevMapping::Local(cats_right) = &**ca_right.get_rev_map() else {
+        unreachable!()
+    };
+
+    let cats_left_hashmap = PlHashMap::from_iter(
+        categories
+            .values_iter()
+            .enumerate()
+            .map(|(k, v)| (v, k as u32)),
+    );
+    let mut new_categories = slots_to_mut(categories);
+    let mut idx_mapping = PlHashMap::with_capacity(cats_right.len());
+
+    for (idx, s) in cats_right.values_iter().enumerate() {
+        if let Some(v) = cats_left_hashmap.get(&s) {
+            idx_mapping.insert(idx as u32, *v);
+        } else {
+            idx_mapping.insert(idx as u32, new_categories.len() as u32);
+            new_categories.push(Some(s));
+        }
     }
+    let new_rev_map = Arc::new(RevMapping::Local(new_categories.into()));
+    Ok((
+        ca_right
+            .physical()
+            .into_iter()
```

Comment:
Can you replace `into_iter` with:


```rust
                self.downcast_iter()
                    .flatten()
```

That ensure we don't go via a `dyn Iterator`.